### PR TITLE
BUG catch more versions

### PR DESCRIPTION
**Pull request by beckermr**
`wget --spider` is failing on this url: https://eups.lsst.codes/stack/src/tags/w_2021_07.list

### CODE REVIEW

**Comment by beckermr** [comment]
This is working fine for openssh and solves my problem. cc @isuruf and thanks for the help!

File: conda_forge_tick/update_sources.py [hunk 1]
```diff
@@ -354,7 +358,13 @@ def url_exists(url: str) -> bool:
...
     if "No such file" in output.decode("utf-8"):
         return False
-    if "not retrieving" in output.decode("utf-8"):
+    if (
+        "not retrieving"
+        in output.decode(
+            "utf-8",
+        )
+        and "Remote file exists" not in output.decode("utf-8")
+    ):
```
Comment (isuruf) [inline conda_forge_tick/update_sources.py:367]: This is equivalent to not having this if condition at all. From my experience, `Remote file exists` doesn't mean anything if not retrieving is in the output.

Comment (beckermr) [reply conda_forge_tick/update_sources.py:367]: Well in my case the file exists but wget refuses to retrieve it:

```
(bot-dev) clarence:cf-scripts beckermr$ wget --spider https://eups.lsst.codes/stack/src/tags/w_2021_07.list
Spider mode enabled. Check if remote file exists.
--2021-02-12 12:04:56--  https://eups.lsst.codes/stack/src/tags/w_2021_07.list
Resolving eups.lsst.codes (eups.lsst.codes)... 104.155.133.93
Connecting to eups.lsst.codes (eups.lsst.codes)|104.155.133.93|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 5414 (5.3K)
Remote file exists and could contain further links,
but recursion is disabled -- not retrieving.

(bot-dev) clarence:cf-scripts beckermr$ wget https://eups.lsst.codes/stack/src/tags/w_2021_07.list
--2021-02-12 12:05:03--  https://eups.lsst.codes/stack/src/tags/w_2021_07.list
Resolving eups.lsst.codes (eups.lsst.codes)... 104.155.133.93
Connecting to eups.lsst.codes (eups.lsst.codes)|104.155.133.93|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 5414 (5.3K)
Saving to: ‘w_2021_07.list’

w_2021_07.list                             100%[======================================================================================>]   5.29K  --.-KB/s    in 0s      

2021-02-12 12:05:03 (10.6 MB/s) - ‘w_2021_07.list’ saved [5414/5414]

```

so your heuristic is wrong here.

Comment (beckermr) [reply conda_forge_tick/update_sources.py:367]: This message looks extremely common so I am inclined to merge this PR.

https://askubuntu.com/questions/389910/get-the-size-of-a-remote-file-from-terminal

Comment (isuruf) [reply conda_forge_tick/update_sources.py:367]: > so your heuristic is wrong here.

Giving one example where a heuristic is wrong and claiming that the heuristic is wrong does not make any sense to me.

Comment (beckermr) [reply conda_forge_tick/update_sources.py:367]: I don't want to debate semantics here. 

My point is simply that the code as written is failing when a url does in fact exist. Do you have an example url where wget returns "Remote file exists and could contain further links,
but recursion is disabled -- not retrieving." in the output and in fact the remote file does not exist?

Comment (beckermr) [reply conda_forge_tick/update_sources.py:367]: also this solves the github behavior:

```
In [3]: r = requests.head("https://github.com/beckermr/pizza-cutter/archive/0.2.5.tar.gz", allow_redirects=True)

In [4]: r
Out[4]: <Response [404]>

In [5]: r = requests.head("https://github.com/beckermr/pizza-cutter/archive/0.2.2.tar.gz", allow_redirects=True)

In [6]: r
Out[6]: <Response [200]>
```

so we should do that.

**Review by isuruf — CHANGES_REQUESTED** [review]
Can you run the new function over all the feedstocks and post which feedstocks/versions are now caught/missed so that we can see if this is beneficial or not?

**Comment by beckermr** [comment]
> Can you run the new function over all the feedstocks and post which feedstocks/versions are now caught/missed so that we can see if this is beneficial or not?

Absolutely not. This function is missing my feedstock and this PR works. It is sending the same head request that wget --spider is sending. There is no reason to do this.

**Comment by beckermr** [comment]
I tested the github urls in the tests and they work fine.

**Comment by isuruf** [comment]
> Absolutely not.

Well then I'm against merging this. You are just ignoring the efforts I put into making sure this works just because you found a false negative whereas I found lots of false positives when writing this code initially.

**Comment by beckermr** [comment]
> You are just ignoring the efforts I put into making sure this works just because you found a false negative whereas I found lots of false positives when writing this code initially.

Please send a list of the false positives and I will test them to make sure they work.

I got the github one. What else did you find?

**Comment by isuruf** [comment]
> Please send a list of the false positives and I will test them to make sure they work.

I don't have them right now and that's why I asked you to run the tests.

**Comment by beckermr** [comment]
> You are just ignoring the efforts

I am not trying to ignore your efforts at all. They are greatly appreciated as are your comments. However since none of the results were documented in a test in any way, it is a big thing to ask someone else to repeat them. If we document what we find so others can look, that will make this a lot easier.

**Comment by beckermr** [comment]
I did some digging here with help from @CJ-Wright . As far as I can tell, I was the one who changed the code to use `--spider` in this PR

https://github.com/regro/cf-scripts/pull/801/files

Previous to this, @isuruf a PR to eliminate false positives. However, those were related to urls with weird versions returning the same old version. See this one

https://github.com/regro/cf-scripts/pull/488

The PR here doesn't change any of the hash testing in the bot now (which was moved to another place) and so should leave that issue fixed. 

@regro/auto-tick Can anyone think of any other things we are missing here?

**Comment by isuruf** [comment]
> As far as I can tell, I was the one who changed the code to use --spider in this PR

I'm sorry that's simply not true. It was added at least a couple of years ago. https://github.com/regro/cf-scripts/pull/92/files

**Comment by beckermr** [comment]
No reason to apologize at all and thank you for the link to your older one! :)

The one I found above was the oldest one I could find and my apologies for the mistake.

Here is the first commit which I found based on the one you sent above: https://github.com/regro/cf-scripts/commit/0e777f8fe8a154aea36d71be5b149d5472595754

**Comment by beckermr** [comment]
`requests` barfs on FTP urls which I had forgotten. I added a test and curl is working fine.

**Comment by CJ-Wright** [comment]
It seems that the `--spider` was not added to weed out false positives, seeing as we ran into issues with false positives well after it was added. My reading of the wget manual is that the spider option allows wget to just check if a url exists. My understanding of this PR is that it replaces the spider functionality, but not the subsequent false positive detection via hash checking.

I'm happy to hear examples to the contrary. It seems that neither of the PRs mentioned previously seem to have tests associated with the issues they seem to solve. However, whatever tests we do have seem to pass now, in addition to the tests that @beckermr added. 

In the absence of these tests and considering that this is only changing the url existence criteria I think it is fair to move forward with this. Should we see false positives to appear as they have in the past we can incorporate them (or something similar to them) into the tests and fix bugs as they arise. Exhaustive testing of the current ecosystem for this seems like it would be difficult and may not yield the results we need. The problems of false positives mentioned above could be transitory or only impact a url that has yet to be added to a feedstock, so such a test would not prove that this approach would work for all time, similar to how it doesn't prove that the current approach would work for all time. In the meantime we do seem to have a bug here with versions not being caught, which this PR seems to solve.

To the earlier points raised about the effort put into the bot. I humbly appreciate all the good faith work put into the bot, of which there has been much. The growth of scope, usage, and user engagement with the bot is always astonishing to me. That the bot operates at it's current level is a testament to the work that all the contributors put in. That being said, I personally put more stock in the bot's successful operation than the authorship of individual lines of code. If there are issues with the approach provided in this PR then we should all attempt to make it better, if there are other options we should consider exploring them. I would really appreciate it if the tone of the conversation could focus more on these aspects.

**Comment by beckermr** [comment]
I want to say for my part that I am very sorry I let my emotions get the best of me. I spoke too strongly and was not considerate. I am sorry for that @isuruf.

**Comment by beckermr** [comment]
I did manage to produce a diff after about 1.5 hours of my laptop running its fan on full blast. :)

Unfortunately due to transient url failures it is nearly impossible to interpret. The new code finds a lot of new versions and by hand checking of where things are missed shows they are failures. However something comprehensive would require several hours more work to check.

**Comment by beckermr** [comment]
Link: https://gist.github.com/beckermr/173f2e235402cd585e201518fd090d53

**Comment by isuruf** [comment]
Thanks. I'll have a look at the diff and report back

**Pull request closed without merging**